src/game/common/rts/scorekeeper.cpp

```diff
@@ -0,0 +1,453 @@
+                        ++m_totalUnitsBuilt;
+                        add_to_map = true;
+                    }
+                }
+            }
+        }
+
+        if (add_to_map) {
+            int count = 0;
+
+            auto iter = m_objectsLost.find(tmpl);
+
+            if (iter != m_objectsLost.end()) {
+                count = (*iter).second;
+            }
+
+            int new_count = count + 1;
+            // needs confirming is the right function
+            m_objectsLost[tmpl] = new_count;
+        }
+    }
+}
+
+void ScoreKeeper::Remove_Object_Built(const Object *object)
+{
+    if (g_theGameLogic->Get_Enable_Scoring()) {
+
+        bool remove_from_map = false;
+
+        const ThingTemplate *tmpl = object->Get_Template();
+
+        if (tmpl->Is_KindOf_Multi(ScoringBuildingMask, KINDOFMASK_NONE)) {
+
+            if (!object->Get_Status(OBJECT_STATUS_UNDER_CONSTRUCTION)) {
+                --m_totalBuildingsBuilt;
+                remove_from_map = true;
+            }
+
+        } else {
+            if (tmpl->Is_KindOf_Multi(ScoringBuildingCreateMask, KINDOFMASK_NONE)) {
+
+                --m_totalBuildingsBuilt;
+                remove_from_map = true;
+
+            } else {
+
+                if (tmpl->Is_KindOf(KINDOF_INFANTRY) || tmpl->Is_KindOf(KINDOF_VEHICLE)) {
+
+                    if (tmpl->Is_KindOf(KINDOF_SCORE) || tmpl->Is_KindOf(KINDOF_SCORE_CREATE)) {
+                        --m_totalUnitsBuilt;
+                        remove_from_map = true;
+                    }
+                }
+            }
+        }
+
+        if (remove_from_map) {
+            int count = 0;
+
+            auto iter = m_objectsLost.find(tmpl);
+
+            if (iter != m_objectsLost.end()) {
+                count = (*iter).second;
+            }
+
+            int new_count = count - 1;
+            // needs confirming is the right function
+            m_objectsLost[tmpl] = new_count;
+        }
+    }
+}
+
+int ScoreKeeper::Get_Total_Units_Built(BitFlags<KINDOF_COUNT> must_be_set, BitFlags<KINDOF_COUNT> must_be_clear)
+{
+    int total = 0;
+
+    for (auto &it : m_objectsBuilt) {
+        if (it.first->Is_KindOf_Multi(must_be_set, must_be_clear)) {
+            total += it.second;
+        }
+    }
+
+    return total;
+}
+
+int ScoreKeeper::Get_Total_Objects_Built(const ThingTemplate *tmpl)
+{
+    int total = 0;
+
+    for (auto &it : m_objectsBuilt) {
+        if (it.first->Is_Equivalent_To(tmpl)) {
+            ++total;
+        }
+    }
+
+    return total;
+}
+
+void ScoreKeeper::Add_Object_Captured(const Object *object)
+{
+    if (g_theGameLogic->Get_Enable_Scoring()) {
+        bool add_to_map = false;
+        const ThingTemplate *tmpl = object->Get_Template();
+
+        if (tmpl->Is_KindOf(KINDOF_STRUCTURE)) {
+
+            if (tmpl->Is_KindOf(KINDOF_SCORE)) {
+                ++int_dwordA0_capturedscore;
+            } else {
+                ++int_dword9C_capturedTech;
```

Comment:
I tested this in game. It triggers when completing capture of civilian building. On some unofficial map it also triggered when beginning to capture a civilian building, which made little sense.